### PR TITLE
correctly type Expr.ofTupleFromSeq for arity > 22

### LINKED ISSUE #17257: asExprOf fails for tuples larger than 22
## Compiler version
3.2.2

## Minimized code

```Scala
// file a.scala
inline def summonAllOptimized[T <: Tuple]: T =
  ${ summonAllOptimizedImpl[T] }

private def summonAllOptimizedImpl[T <: Tuple: Type](using q: Quotes): Expr[T] = {
  import q.reflect.*

  Expr
    .ofTupleFromSeq(typesOfTuple(TypeRepr.of[T], Nil).map { tpe =>
      tpe.asType match {
        case '[t] =>
          Expr.summon[t].getOrElse(report.errorAndAbort(s"Unable to to find implicit instance for ${tpe.show}"))
      }
    })
    .asExprOf[T]
}

@tailrec
private[derivation] def typesOfTuple(
    using q: Quotes
)(tpe: q.reflect.TypeRepr, acc: List[q.reflect.TypeRepr]): List[q.reflect.TypeRepr] =
  import q.reflect.*
  val cons = Symbol.classSymbol("scala.*:")
  tpe.widenTermRefByName.dealias match
    case AppliedType(fn, tpes) if defn.isTupleClass(fn.typeSymbol) =>
      tpes.reverse_:::(acc)
    case AppliedType(tp, List(headType, tailType)) if tp.derivesFrom(cons) =>
      typesOfTuple(tailType, headType :: acc)
    case tpe =>
      if tpe.derivesFrom(Symbol.classSymbol("scala.EmptyTuple")) then acc.reverse
      else report.errorAndAbort(s"Unknown type encountered in tuple ${tpe.show}")

//file b.scala
val test = Helpers.summonAllOptimized[(
    ValueOf["a"], ValueOf["a"], ValueOf["a"], ValueOf["a"], ValueOf["a"],
    ValueOf["a"], ValueOf["a"], ValueOf["a"], ValueOf["a"], ValueOf["a"],
    ValueOf["a"], ValueOf["a"], ValueOf["a"], ValueOf["a"], ValueOf["a"],
    ValueOf["a"], ValueOf["a"], ValueOf["a"], ValueOf["a"], ValueOf["a"],
    ValueOf["a"], ValueOf["a"], ValueOf["a"] //Commenting out the last one here fixes the compile error
  )]
```

## Output

```scala
[error]     |Exception occurred while executing macro expansion.
[error]     |java.lang.Exception: Expr cast exception: scala.Tuple.fromIArray[scala.Any](scala.IArray$package.IArray.apply[scala.Any](new scala.ValueOf["a"]("a"), new scala.ValueOf["a"]("a"), new scala.ValueOf["a"]("a"), new scala.ValueOf["a"]("a"), new scala.ValueOf["a"]("a"), new scala.ValueOf["a"]("a"), new scala.ValueOf["a"]("a"), new scala.ValueOf["a"]("a"), new scala.ValueOf["a"]("a"), new scala.ValueOf["a"]("a"), new scala.ValueOf["a"]("a"), new scala.ValueOf["a"]("a"), new scala.ValueOf["a"]("a"), new scala.ValueOf["a"]("a"), new scala.ValueOf["a"]("a"), new scala.ValueOf["a"]("a"), new scala.ValueOf["a"]("a"), new scala.ValueOf["a"]("a"), new scala.ValueOf["a"]("a"), new scala.ValueOf["a"]("a"), new scala.ValueOf["a"]("a"), new scala.ValueOf["a"]("a"), new scala.ValueOf["a"]("a"))(scala.reflect.ClassTag.Any))
[error]     |of type: scala.Tuple
[error]     |did not conform to type: scala.*:[scala.ValueOf["a"], scala.*:[scala.ValueOf["a"], scala.*:[scala.ValueOf["a"], scala.*:[scala.ValueOf["a"], scala.*:[scala.ValueOf["a"], scala.*:[scala.ValueOf["a"], scala.*:[scala.ValueOf["a"], scala.*:[scala.ValueOf["a"], scala.*:[scala.ValueOf["a"], scala.*:[scala.ValueOf["a"], scala.*:[scala.ValueOf["a"], scala.*:[scala.ValueOf["a"], scala.*:[scala.ValueOf["a"], scala.*:[scala.ValueOf["a"], scala.*:[scala.ValueOf["a"], scala.*:[scala.ValueOf["a"], scala.*:[scala.ValueOf["a"], scala.*:[scala.ValueOf["a"], scala.*:[scala.ValueOf["a"], scala.*:[scala.ValueOf["a"], scala.*:[scala.ValueOf["a"], scala.*:[scala.ValueOf["a"], scala.*:[scala.ValueOf["a"], scala.Tuple$package.EmptyTuple]]]]]]]]]]]]]]]]]]]]]]]
[error]     |
[error]     |   at scala.quoted.runtime.impl.QuotesImpl.asExprOf(QuotesImpl.scala:73)
[error]     |   at perspective.derivation.Helpers$.summonAllOptimizedImpl(Helpers.scala:157)
[error]     |   at perspective.derivation.Helpers$.inline$summonAllOptimizedImpl(Helpers.scala:147)
```

## Expectatio
I would expect the above to work just like it works for tuples of size 22 or less, or that this was at least better communicated.

### PR DESCRIPTION
**Pull request by bishabosha**
identified that for all `Seq` with arity <= 22, the result had the most precise type possible, for > 22 then it was only `Tuple`

fixes #17257

### CODE REVIEW

**Review by dwijnand — APPROVED** [review]

**Comment by bishabosha** [comment]
Also seeing the same failure here: `tests/neg-custom-args/fatal-warnings/i15503g.scala` it was failing #16056 too. Does anyone know what it is?

**Comment by dwijnand** [comment]
Looks like a semantic merge conflict on the Wunused merges?  @szymon-rd wdyt?

**Comment by dwijnand** [comment]
Ah yeah, https://github.com/lampepfl/dotty/pull/17263

**Comment by nicolasstucki** [comment]
@bishabosha also take into account https://github.com/lampepfl/dotty/issues/17257#issuecomment-1510947785

**Comment by bishabosha** [comment]
@nicolasstucki I added a test for `Expr.ofTuple` with 23 elems

**Review by nicolasstucki — APPROVED** [review]

**Comment by bishabosha** [comment]
could this be backported for 3.3.1 ?

File: library/src/scala/quoted/Expr.scala [hunk 1]
```diff
@@ -214,6 +214,18 @@ object Expr {
...
+  private def ofTupleFromSeqXXL(seq: Seq[Expr[Any]])(using Quotes): Expr[Tuple] =
+    val tupleTpe = tupleTypeFromSeq(seq)
+    tupleTpe.asType match
+      case '[tpe] =>
+        '{ Tuple.fromIArray(IArray(${Varargs(seq)}*)).asInstanceOf[tpe & Tuple] }
+
+  private def tupleTypeFromSeq(seq: Seq[Expr[Any]])(using Quotes): quotes.reflect.TypeRepr =
+    import quotes.reflect.*
+    val consRef = Symbol.classSymbol("scala.*:").typeRef
+    seq.foldLeft(TypeRepr.of[EmptyTuple]) { (ts, expr) =>
```
Comment (xuwei-k) [inline library/src/scala/quoted/Expr.scala:226]: - https://github.com/scala/scala3/issues/21779

Comment (joroKr21) [reply library/src/scala/quoted/Expr.scala:226]: https://github.com/scala/scala3/pull/21782

**Pull request merged**